### PR TITLE
Enable and make stage0 landing pads optional

### PR DESCRIPTION
**Pull request by nagisa**
Part of #28710

Landing pads during stage0 are now enabled by defaullt. Since this has its downsides and upsides either way, I made it possible to change the option through configure.

### CODE REVIEW

**Comment by alexcrichton** [comment]
cc @arielb1, thoughts?

File: configure [hunk 1]
```diff
@@ -578,6 +578,7 @@ opt ccache 0 "invoke gcc/clang via ccache to reuse object files between builds"
 opt local-rust 0 "use an installed rustc rather than downloading a snapshot"
 opt llvm-static-stdcpp 0 "statically link to libstdc++ for LLVM"
 opt rpath 0 "build rpaths into rustc itself"
+opt landing-pads 1 "enable landing pads during bootstrap"
```
Comment (alexcrichton) [inline configure]: Perhaps this could be called `stage0-landing-pads` to ensure it's not mistaken for disabling landing pads entirely?

**Comment by nagisa** [comment]
Whatever way this PR goes, IMHO it still would be great to fix compiletest to work with or without landing-pads.

File: mk/main.mk [hunk 2]
```diff
@@ -170,7 +170,10 @@ RUST_LIB_FLAGS_ST3 += -C prefer-dynamic
 
 # Landing pads require a lot of codegen. We can get through bootstrapping faster
 # by not emitting them.
-RUSTFLAGS_STAGE0 += -Z no-landing-pads
+
+ifdef CFG_DISABLE_STAGE1_LANDING_PADS
```
Comment (alexcrichton) [inline mk/main.mk]: Could this be called stage0 to mirror the option it's used for below? I guess it kinda depends on what a "stage1 bootstrap" means, which also depends on who you talk to! May be worth just keeping these numbers in sync though to alleviate confusion of the off-by-one nature?

**Comment by alexcrichton** [comment]
> IMHO it still would be great to fix compiletest to work with or without landing-pads

Certainly!

**Comment by alexcrichton** [comment]
@bors: r+ cefadf05f9319f01b0245de72f4e5941a046b0e9

Thanks!

**Comment by arielb1** [comment]
@alexcrichton 

This is a *decent* interim solution, but I would prefer a real one.

**Comment by nagisa** [comment]
I got no way to investigate this failure, sadly.

I could make pads disabled by default, but that makes this PR pretty useless IMO.

**Comment by alexcrichton** [comment]
I think that unwinding for MSVC 64-bit landed sometime after the last snapshot causing these problems (e.g. MSVC 64-bit in stage0 wants the old method of unwinding). If you tweak [these cfgs](https://github.com/rust-lang/rust/blob/17a2cb443a9dda03f2c35853fb69b981f179a4df/src/libstd/sys/common/unwind/mod.rs#L79-L87) to have 64-bit MSVC in stage0 routed to `seh.rs` and only routed to `seh64_gnu.rs` in stage1+ I think it'll solve that error

**Comment by nagisa** [comment]
@alexcrichton can we get new snapshots instead?

**Comment by alexcrichton** [comment]
@bors: r+ 8cf5cf0

If we can avoid new snapshots for now it'd probably be good to keep doing so, but if we really need to we could surely make new snapshots!

**Comment by nagisa** [comment]
Well, at least MSVC didn’t fail, but it seems wiser to wait for snaps rather than trying to fix all windows’ landing pad implementations for stage0.

**Comment by alexcrichton** [comment]
Oh right, I think in stage0 x86_64-pc-windows-gnu needs to be routed to `gcc.rs` instead of `seh64_gnu.rs` as well. I think that'll do the trick though?

**Comment by nagisa** [comment]
OK, lets retry this.

**Comment by alexcrichton** [comment]
Looks like the travis failure may be legit

**Comment by alexcrichton** [comment]
Aha, looks like stage0 64-bit gnu windows needs `seh64_gnu.rs` (support had landed at that time), it must have just been after the snapshot that support was added for MSVC. Routing it to that unwinding strategy gets the bootstrap farther into stage0 for me.

**Pull request merged**